src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -264,6 +282,68 @@ private static MethodInfo GetMethodInfo<T>(Expression<T> expr)
             return Expression.New(constructor, constructorArguments);
         }
 
+#if NETSTANDARD2_1_OR_GREATER || NET6_0_OR_GREATER
+        private static ObjectFactory CreateFactoryReflection(
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type instanceType,
+            Type?[] argumentTypes)
+        {
+            FindApplicableConstructor(instanceType, argumentTypes, out ConstructorInfo constructor, out int?[] parameterMap);
+
+            ParameterInfo[]? constructorParameters = constructor.GetParameters();
+            List<FactoryParameterContext> parameters = new List<FactoryParameterContext>();
+            for (int i = 0; i < constructorParameters.Length; i++)
+            {
+                ParameterInfo constructorParameter = constructorParameters[i];
+                bool hasDefaultValue = ParameterDefaultValue.TryGetDefaultValue(constructorParameter, out object? defaultValue);
+
+                parameters.Add(new FactoryParameterContext(constructorParameter, hasDefaultValue, defaultValue, parameterMap[i]));
+            }
+
+            return (IServiceProvider serviceProvider, object?[]? arguments) =>
+            {
+                var constructorArguments = new object?[parameters.Count];
+                for (int i = 0; i < parameters.Count; i++)
+                {
+                    FactoryParameterContext parameter = parameters[i];
+                    if (parameter.ArgumentIndex is { } index)
+                    {
+                        constructorArguments[i] = arguments?[index];
+                    }
+                    else
+                    {
+                        constructorArguments[i] = GetService(
+                            serviceProvider,
+                            parameter.ParameterInfo.ParameterType,
```

Comment:
This is the only use of `ParameterInfo`. Cache `ParameterType` directly to avoid this virtual call?